pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+		return fmt.Errorf("invalid node count: %d. Must be at least 1", spec.Storage.NodeCount)
+	}
+
+	if err := validatePercentValue(spec.CachePercent, "cache"); err != nil {
+		return err
+	}
+	if err := validatePercentValue(spec.MaxSQLMemoryPercent, "maxSQLMemory"); err != nil {
+		return err
+	}
+
+	if _, _, err := getPortsFromSpec(spec.Network); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validatePercentValue(value int, name string) error {
+	if value < 0 || value > 100 {
+		return fmt.Errorf("invalid value (%d) for %s percent, must be between 0 and 100 inclusive", value, name)
+	}
+
+	return nil
+}
+
+func createAppLabels() map[string]string {
+	return map[string]string{
+		k8sutil.AppAttr: appName,
+	}
+}
+
+func createServicePorts(httpPort, grpcPort int32) []v1.ServicePort {
+	return []v1.ServicePort{
+		{
+			// The main port, served by gRPC, serves Postgres-flavor SQL, internode traffic and the cli.
+			Name:       grpcPortName,
+			Port:       int32(grpcPort),
+			TargetPort: intstr.FromInt(int(grpcPort)),
+		},
+		{
+			// The secondary port serves the UI as well as health and debug endpoints.
+			Name:       httpPortName,
+			Port:       int32(httpPort),
+			TargetPort: intstr.FromInt(int(httpPort)),
+		},
+	}
+}
+
+func getPortsFromSpec(networkSpec rookv1alpha2.NetworkSpec) (httpPort, grpcPort int32, err error) {
+	for _, p := range networkSpec.Ports {
+		switch p.Name {
+		case httpPortName:
+			httpPort = p.Port
+		case grpcPortName:
+			grpcPort = p.Port
+		default:
+			return 0, 0, fmt.Errorf("unknown port name: %s", p.Name)
+		}
+	}
+
+	if httpPort == 0 {
+		httpPort = httpPortDefault
+	}
+	if grpcPort == 0 {
+		grpcPort = grpcPortDefault
+	}
+
+	return httpPort, grpcPort, nil
+}
+
+// creates a qualified name of the replica service for a given replica and namespace, e.g., cockroachdb-0.cockroachdb.rook-cockroachdb
+func createQualifiedReplicaServiceName(replicaNum int, namespace string) string {
+	return fmt.Sprintf("%s-%d.%s.%s", appName, replicaNum, appName, namespace)
+}
+
+func createCommand(cluster *cluster) string {
+	var insecureFlag string
+	if !cluster.spec.Secure {
+		insecureFlag = "--insecure"
+	}
+
+	var joinFlag string
+	if cluster.spec.Storage.NodeCount > 0 {
+		// generate a list of DNS names of instances to join with that takes into account the service name of each stateful set
+		// instance and the namespace they are in. e.g., cockroachdb-0.cockroachdb.rook-cockroachdb
+		joinList := make([]string, cluster.spec.Storage.NodeCount)
+		for i := 0; i < cluster.spec.Storage.NodeCount; i++ {
+			joinList[i] = createQualifiedReplicaServiceName(i, cluster.namespace)
+		}
+
+		joinFlag = strings.Join(joinList, ",")
+	}
+
+	// The use of qualified `hostname -f` is crucial: Other nodes aren't able to look up the unqualified hostname.
+	return fmt.Sprintf("exec /cockroach/cockroach start --logtostderr %s --advertise-host $(hostname -f) --http-host 0.0.0.0 --join %s --cache %s%% --max-sql-memory %s%%",
```

Comment:
do you need the `exec` command? I thought our `executor` had a method that started a process without waiting

Reply:
This came from the existing k8s solution for cockroach, and it is used directly as the command/entrypoint to the cockroachdb pods.  It will not touch our executor at all.  That being said, it may very well not be needed, I'll try to test without it.